cassovary-core/src/main/scala/com/twitter/cassovary/graph/SharedArrayBasedDirectedGraph.scala

```diff
@@ -42,237 +42,320 @@ object SharedArrayBasedDirectedGraph {
+                nodeInShard =>
+                  id = nodeInShard * numOfShards + shard
+                  if (nodeIdSet(id) == 1) {
+                    lengthTable(id) = if (inEdgesSizes(id) != null) inEdgesSizes(id).get else 0
+                    if (lengthTable(id) > 0)
+                      offsetTable(id) = nextFreeCellPerShard(shard).getAndAdd(lengthTable(id))
+                    inEdgesSizes(id).set(0)
+                  }
               }
             }
           }
+          Future.join(futures)
         }
-
-        ExecutorUtils.parallelForEach[Int, Unit](
-            executorService, (0 until numOfShards), instantiateInEdgesTask)
       }
 
-      log.info("populate in edges")
-      statsReceiver.time("graph_load_read_in_edge_from_dump_files") {
-        def readInEdges = {
-          (shard: Int) => {
-            var id, offset, length, edgeIndex, index = 0
-            for(i <- 0 to (maxNodeId / numOfShards)) {
-              id = i * numOfShards + shard
-              if ((id <= maxNodeId) && (nodeIdSet(id) == 1)) {
-                offset = offsetTable(id)
-                length = lengthTable(id)
-                for (edgeOffset <- offset until (offset + length)) {
-                  edgeIndex = sharedEdgeArray(id % numOfShards)(edgeOffset)
-                  index = inEdgesSizes(edgeIndex).getAndIncrement()
-                  reverseEdges(edgeIndex)(index) = id
-                }
+      /**
+       * Fills in edges assuming that lengthTable and offsetTable are filled and
+       * nextFreeCell is 0 everywhere.
+       */
+      def fillInEdges(sharedInEdgesArray: Array[Array[Int]], offsetTable: Array[Int],
+                      lengthTable: Array[Int], nextFreeCellPerNode: Array[AtomicInteger]):
+      Future[Unit] = {
+        log.info("filling in edges")
+        statsReceiver.timeFuture("graph_load_fill_in_edges") {
+          val futures = iterableSeq.map {
+            nodesMaxIdsSeq => futurePool {
+              nodesMaxIdsSeq.map {
+                item =>
+                  item.edges.foreach {
+                    target =>
+                      sharedInEdgesArray(target % numOfShards)(offsetTable(target) +
+                        nextFreeCellPerNode(target).getAndIncrement) = item.id
+                  }
               }
             }
           }
+          Future.join(futures)
         }
-        ExecutorUtils.parallelForEach[Int, Unit](executorService, (0 until numOfShards), readInEdges)
       }
-      Some(reverseEdges)
-    } else None
 
-    new SharedArrayBasedDirectedGraph(nodeIdSet, offsetTable, lengthTable, sharedEdgeArray,
-      reverseDirEdgeArray, maxNodeId, nodeWithOutEdgesMaxId, nodeWithOutEdgesCount,
-      numNodes, numEdges, storedGraphDir)
+      for {
+        (inEdgesSizes, inEdgesShardsSizes) <- findInEdgesAndShardsSizes(sharedEdgeArray, shardsLengths, nodeIdSet)
+        sharedInEdges = instantiateSharedArray(maxNodeId, inEdgesShardsSizes)
```

Comment:
note that maxNodeId should change now as well as it should include nodes that only have in edges, so we need to calculate this by looking at each edge (and not just each node).